### PR TITLE
[Snappi] Adding Many To One Oversubscribe Lossless Lossy Testcase

### PR DESCRIPTION
**Pull request by dks0692**
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/sonic-net/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
Tests the response of lossy and lossless streams destined to a single egress port, with both lossless & lossy traffic causing oversubscription of the egress port. 
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->

Summary:
Fixes # (issue)

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)


### Back port request
- [ ] 201911
- [ ] 202012
- [ ] 202205

### Approach
#### What is the motivation for this PR?
To check if packet drops really happen when lossy traffic suffers from congestion without triggering PFC and WRR scheduler works as expected
#### How did you do it?
It was tested by oversubscribing the rx port by increasing the flow rate of multiple traffic flows and checking for loss one one of the background flow with 40% line rate
#### How did you verify/test it?
Tested against voq chassis with SONiC installed in it


#### Any platform specific information?

#### Supported testbed topology if it's a new test case?
T2, T1 & T0
### Documentation

Testplan - https://github.com/sonic-net/sonic-mgmt/pull/8373
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->
### Output

AzDevOps@6105354dad99:~/sonic-mgmt/tests$ py.test --inventory ../ansible/snappi-sonic --host-pattern sonic-s6100-dut --testbed vms-snappi-sonic --testbed_file ../ansible/testbed.csv --show-capture=stdout --log-cli-level info --showlocals -ra --allow_recover --skip_sanity --disable_loganalyzer snappi_tests/multidut/pfc/test_m2o_oversubscribe_lossless_lossy.py 
/usr/local/lib/python2.7/dist-packages/ansible/parsing/vault/__init__.py:44: CryptographyDeprecationWarning: Python 2 is no longer supported by the Python core team. Support for it is now deprecated in cryptography, and will be removed in the next release.
  from cryptography.exceptions import InvalidSignature
======================================================================================================= test session starts =======================================================================================================
platform linux2 -- Python 2.7.18, pytest-4.6.11, py-1.11.0, pluggy-0.13.1
ansible: 2.8.20
rootdir: /var/AzDevOps/sonic-mgmt/tests, inifile: pytest.ini
plugins: allure-pytest-2.8.22, celery-4.4.7, ansible-2.2.4, forked-1.3.0, html-1.22.1, metadata-1.11.0, repeat-0.9.1, xdist-1.28.0
collecting 1 item                                                                                                                                                                                                                 
------------------------------------------------------------------------------------------------------- live log collection -------------------------------------------------------------------------------------------------------
18:23:32 __init__.pytest_collection_modifyitems   L0582 INFO   | Available basic facts that can be used in conditional skip:
{
  "topo_type": "ptf", 
  "testbed": "vms-snappi-sonic", 
  "topo_name": "ptf64"
}

--------------------------------------------------------------------------------------------------------- live log setup ----------------------------------------------------------------------------------------------------------
18:23:32 __init__.set_default                     L0054 INFO   | Completeness level not set during test execution. Setting to default level: CompletenessLevel.basic
18:23:32 __init__.check_test_completeness         L0152 INFO   | Test has no defined levels. Continue without test completeness checks
18:23:39 ptfhost_utils.run_icmp_responder_session L0239 INFO   | Skip running icmp_responder at session level, it is only for dualtor testbed with active-active mux ports.
18:23:39 conftest.creds_on_dut                    L0704 INFO   | dut sonic-s6100-dut belongs to groups [u'snappi-sonic', u'sonic', u'sonic_dell64_40', 'fanout']
18:23:39 conftest.creds_on_dut                    L0728 INFO   | skip empty var file ../ansible/group_vars/all/corefile_uploader.yml
18:23:40 conftest.creds_on_dut                    L0728 INFO   | skip empty var file ../ansible/group_vars/all/env.yml
18:23:40 conftest.nbrhosts                        L0521 INFO   | No VMs exist for this topology: ptf64
18:23:40 conftest.core_dump_and_config_check      L2032 INFO   | Collecting core dumps before test on sonic-s6100-dut
18:23:40 conftest.core_dump_and_config_check      L2041 INFO   | Collecting running config before test on sonic-s6100-dut
18:23:45 __init__.sanity_check                    L0125 INFO   | Skip sanity check according to command line argument
18:23:45 conftest.generate_params_dut_hostname    L1108 INFO   | Using DUTs ['sonic-s6100-dut'] in testbed 'vms-snappi-sonic'
18:23:45 conftest.rand_one_dut_hostname           L0378 INFO   | Randomly select dut sonic-s6100-dut for testing
18:23:45 __init__._fixture_generator_decorator    L0081 INFO   | -------------------- fixture enable_packet_aging_after_test setup starts --------------------
18:23:45 __init__._fixture_generator_decorator    L0085 INFO   | -------------------- fixture enable_packet_aging_after_test setup ends --------------------
18:23:45 __init__._fixture_generator_decorator    L0081 INFO   | -------------------- fixture rand_lossless_prio setup starts --------------------
18:23:45 __init__._fixture_generator_decorator    L0085 INFO   | -------------------- fixture rand_lossless_prio setup ends --------------------
18:23:45 __init__._fixture_generator_decorator    L0081 INFO   | -------------------- fixture rand_lossy_prio setup starts --------------------
18:23:45 __init__._fixture_generator_decorator    L0085 INFO   | -------------------- fixture rand_lossy_prio setup ends --------------------
18:23:45 __init__._fixture_generator_decorator    L0081 INFO   | -------------------- fixture start_pfcwd_after_test setup starts --------------------
18:23:45 __init__._fixture_generator_decorator    L0085 INFO   | -------------------- fixture start_pfcwd_after_test setup ends --------------------
18:23:45 __init__._fixture_func_decorator         L0069 INFO   | -------------------- fixture snappi_api_serv_ip setup starts --------------------
18:23:45 __init__._fixture_func_decorator         L0076 INFO   | -------------------- fixture snappi_api_serv_ip setup ends --------------------
18:23:45 __init__._fixture_func_decorator         L0069 INFO   | -------------------- fixture snappi_api_serv_port setup starts --------------------
18:23:45 __init__._fixture_func_decorator         L0076 INFO   | -------------------- fixture snappi_api_serv_port setup ends --------------------
18:23:45 __init__._fixture_generator_decorator    L0081 INFO   | -------------------- fixture snappi_api setup starts --------------------
18:23:45 __init__._fixture_generator_decorator    L0085 INFO   | -------------------- fixture snappi_api setup ends --------------------
18:23:45 __init__._fixture_func_decorator         L0069 INFO   | -------------------- fixture prio_dscp_map setup starts --------------------
18:23:47 __init__._fixture_func_decorator         L0076 INFO   | -------------------- fixture prio_dscp_map setup ends --------------------
18:23:47 __init__._fixture_func_decorator         L0069 INFO   | -------------------- fixture lossless_prio_list setup starts --------------------
18:23:48 __init__._fixture_func_decorator         L0076 INFO   | -------------------- fixture lossless_prio_list setup ends --------------------
18:23:48 __init__.loganalyzer                     L0045 INFO   | Log analyzer is disabled
18:23:48 __init__._fixture_func_decorator         L0069 INFO   | -------------------- fixture get_multidut_snappi_ports setup starts --------------------
18:23:48 __init__._fixture_func_decorator         L0076 INFO   | -------------------- fixture get_multidut_snappi_ports setup ends --------------------
---------------------------------------------------------------------------------------------------------- live log call ----------------------------------------------------------------------------------------------------------
18:23:48 snappi_fixtures.__intf_config_multidut   L0732 INFO   | Configuring Dut: sonic-s6100-dut with port Ethernet12 with IP 20.0.1.1/24
18:23:49 snappi_fixtures.__intf_config_multidut   L0732 INFO   | Configuring Dut: sonic-s6100-dut with port Ethernet8 with IP 20.0.2.1/24
18:23:51 snappi_fixtures.__intf_config_multidut   L0732 INFO   | Configuring Dut: sonic-s6100-dut with port Ethernet0 with IP 20.0.3.1/24
18:23:58 connection._warn                         L0246 WARNING| Verification of certificates is disabled
18:23:58 connection._info                         L0243 INFO   | Determining the platform and rest_port using the 10.36.78.135 address...
18:23:58 connection._warn                         L0246 WARNING| Unable to connect to http://10.36.78.135:443.
18:23:58 connection._info                         L0243 INFO   | Connection established to `https://10.36.78.135:443 on linux`
18:24:11 connection._info                         L0243 INFO   | Using IxNetwork api server version 9.30.2212.7
18:24:11 connection._info                         L0243 INFO   | User info IxNetwork/ixnetworkweb/admin-30-32270
18:24:12 snappi_api.info                          L1132 INFO   | snappi-0.9.1
18:24:12 snappi_api.info                          L1132 INFO   | snappi_ixnetwork-0.9.1
18:24:12 snappi_api.info                          L1132 INFO   | ixnetwork_restpy-1.0.64
18:24:12 snappi_api.info                          L1132 INFO   | Config validation 0.011s
18:24:15 snappi_api.info                          L1132 INFO   | Ports configuration 1.733s
18:24:15 snappi_api.info                          L1132 INFO   | Captures configuration 0.159s
18:24:18 snappi_api.info                          L1132 INFO   | Add location hosts [10.36.78.53] 2.366s
18:24:22 snappi_api.info                          L1132 INFO   | Location hosts ready [10.36.78.53] 4.166s
18:24:22 snappi_api.info                          L1132 INFO   | Speed conversion is not require for (port.name, speed) : [('Port 0', u'novusHundredGigNonFanOut'), ('Port 1', u'novusHundredGigNonFanOut'), ('Port 2', u'novusHundredGigNonFanOut')]
18:24:22 snappi_api.info                          L1132 INFO   | Aggregation mode speed change 0.436s
18:24:29 snappi_api.info                          L1132 INFO   | Location preemption [10.36.78.53;6;4, 10.36.78.53;6;3, 10.36.78.53;6;1] 0.112s
18:24:54 snappi_api.info                          L1132 INFO   | Location connect [Port 0, Port 1, Port 2] 25.055s
18:24:54 snappi_api.warning                       L1138 WARNING| Port 0 connectedLinkDown
18:24:54 snappi_api.warning                       L1138 WARNING| Port 1 connectedLinkDown
18:24:54 snappi_api.warning                       L1138 WARNING| Port 2 connectedLinkDown
18:24:54 snappi_api.info                          L1132 INFO   | Location state check [Port 0, Port 1, Port 2] 0.251s
18:24:54 snappi_api.info                          L1132 INFO   | Location configuration 39.362s
18:25:10 snappi_api.info                          L1132 INFO   | Layer1 configuration 15.757s
18:25:10 snappi_api.info                          L1132 INFO   | Lag Configuration 0.083s
18:25:10 snappi_api.info                          L1132 INFO   | Convert device config : 0.232s
18:25:10 snappi_api.info                          L1132 INFO   | Create IxNetwork device config : 0.000s
18:25:11 snappi_api.info                          L1132 INFO   | Push IxNetwork device config : 0.655s
18:25:11 snappi_api.info                          L1132 INFO   | Devices configuration 0.963s
18:25:15 snappi_api.info                          L1132 INFO   | Flows configuration 3.922s
18:25:23 snappi_api.info                          L1132 INFO   | Start interfaces 8.202s
18:25:24 snappi_api.info                          L1132 INFO   | IxNet - The Traffic Item was modified. Please perform a Traffic Generate to update the associated traffic Flow Groups
18:25:24 traffic_generation.run_traffic           L0548 INFO   | Wait for Arp to Resolve ...
18:25:25 traffic_generation.run_traffic           L0565 INFO   | Starting transmit on all flows ...
18:25:29 snappi_api.info                          L1132 INFO   | Flows generate/apply 3.095s
18:25:41 snappi_api.info                          L1132 INFO   | Flows clear statistics 11.974s
18:25:41 snappi_api.info                          L1132 INFO   | Captures start 0.000s
18:25:45 snappi_api.info                          L1132 INFO   | Flows start 4.080s
18:25:45 traffic_generation.run_traffic           L0572 INFO   | Polling DUT for traffic statistics for 30 seconds ...
18:26:15 traffic_generation.run_traffic           L0590 INFO   | DUT polling complete
18:26:22 traffic_generation.run_traffic           L0606 INFO   | All test and background traffic flows stopped
18:26:24 traffic_generation.run_traffic           L0629 INFO   | Dumping per-flow statistics
18:26:25 traffic_generation.run_traffic           L0633 INFO   | Stopping transmit on all remaining flows
18:26:25 traffic_generation.verify_m2o_results    L1045 INFO   | Test Flow 2 -> 0 Rate:20, TX Frames:47892720, RX Frames:47892720
18:26:25 traffic_generation.verify_m2o_results    L1045 INFO   | Background Flow 1 -> 0 Rate:20, TX Frames:47892720, RX Frames:47892720
18:26:25 traffic_generation.verify_m2o_results    L1045 INFO   | Test Flow 1 -> 0 Rate:40, TX Frames:95785440, RX Frames:95785440
18:26:25 traffic_generation.verify_m2o_results    L1045 INFO   | Test Flow 2 -> 0 Rate:20, TX Frames:47892720, RX Frames:47892720
18:26:25 traffic_generation.verify_m2o_results    L1052 INFO   | Background Flow 2 -> 0 Rate:40, TX Frames:95785440, RX Frames:83817107
18:26:25 snappi_fixtures.cleanup_config           L0885 INFO   | Removing Configuration on Dut: sonic-s6100-dut with port Ethernet12 with ip :20.0.1.1/24
18:26:27 snappi_fixtures.cleanup_config           L0885 INFO   | Removing Configuration on Dut: sonic-s6100-dut with port Ethernet8 with ip :20.0.2.1/24
18:26:29 snappi_fixtures.cleanup_config           L0885 INFO   | Removing Configuration on Dut: sonic-s6100-dut with port Ethernet0 with ip :20.0.3.1/24
PASSED                                                                                                                                                                                                                                    [100%]
--------------------------------------------------------------------------------------------------------------- live log teardown ---------------------------------------------------------------------------------------------------------------
18:26:31 __init__._fixture_generator_decorator    L0093 INFO   | -------------------- fixture snappi_api teardown starts --------------------
18:26:44 __init__._fixture_generator_decorator    L0102 INFO   | -------------------- fixture snappi_api teardown ends --------------------
18:26:44 __init__._fixture_generator_decorator    L0093 INFO   | -------------------- fixture start_pfcwd_after_test teardown starts --------------------
18:26:45 __init__._fixture_generator_decorator    L0102 INFO   | -------------------- fixture start_pfcwd_after_test teardown ends --------------------
18:26:45 __init__._fixture_generator_decorator    L0093 INFO   | -------------------- fixture rand_lossy_prio teardown starts --------------------
18:26:45 __init__._fixture_generator_decorator    L0102 INFO   | -------------------- fixture rand_lossy_prio teardown ends --------------------
18:26:45 __init__._fixture_generator_decorator    L0093 INFO   | -------------------- fixture rand_lossless_prio teardown starts --------------------
18:26:45 __init__._fixture_generator_decorator    L0102 INFO   | -------------------- fixture rand_lossless_prio teardown ends --------------------
18:26:45 __init__._fixture_generator_decorator    L0093 INFO   | -------------------- fixture enable_packet_aging_after_test teardown starts --------------------
18:26:46 __init__._fixture_generator_decorator    L0102 INFO   | -------------------- fixture enable_packet_aging_after_test teardown ends --------------------
18:26:47 conftest.core_dump_and_config_check      L2069 INFO   | Collecting core dumps after test on sonic-s6100-dut
18:26:48 conftest.core_dump_and_config_check      L2086 INFO   | Collecting running config after test on sonic-s6100-dut
18:26:49 conftest.core_dump_and_config_check      L2211 WARNING| Core dump or config check failed for test_m2o_oversubscribe_lossless_lossy.py, results: {"core_dump_check": {"new_core_dumps": {"sonic-s6100-dut": []}, "pass": true}, "config_db_check": {"cur_only_config": {"sonic-s6100-dut": {"null": {"INTERFACE": {"Ethernet4|20.0.3.1/24": {}, "Ethernet8": {}, "Ethernet4|20.0.2.1/24": {}, "Ethernet4": {}, "Ethernet8|20.0.1.1/24": {}, "Ethernet12|20.0.2.1/24": {}, "Ethernet8|20.0.3.1/24": {}, "Ethernet12": {}}}}}, "inconsistent_config": {"sonic-s6100-dut": {"null": {}}}, "pre_only_config": {"sonic-s6100-dut": {"null": {"VLAN_MEMBER": {"Vlan2|Ethernet8": {"tagging_mode": "untagged"}, "Vlan2|Ethernet0": {"tagging_mode": "untagged"}, "Vlan2|Ethernet12": {"tagging_mode": "untagged"}, "Vlan2|Ethernet4": {"tagging_mode": "untagged"}}}}}, "pass": false}}


=============================================================================================================== warnings summary ================================================================================================================
/usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:545
  /usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:545: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: tests.common.plugins.loganalyzer
    self.import_plugin(import_spec)

/usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:545
  /usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:545: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: tests.common.plugins.sanity_check
    self.import_plugin(import_spec)

/usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:545
  /usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:545: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: tests.common.dualtor
    self.import_plugin(import_spec)

snappi_tests/dinesh_response_time_pr/test_m2o_oversubscribe_lossless_lossy.py::test_m2o_oversubscribe_lossless_lossy[linecard_configuration_set0-non_chassis_single_line_card]
  /usr/local/lib/python2.7/dist-packages/pytest_ansible/module_dispatcher/v28.py:81: UserWarning: provided hosts list is empty, only localhost is available
    warnings.warn("provided hosts list is empty, only localhost is available")

-- Docs: https://docs.pytest.org/en/latest/warnings.html
------------------------------------------------------------------------------------------------------------ live log sessionfinish -------------------------------------------------------------------------------------------------------------
18:26:49 __init__.pytest_terminal_summary         L0064 INFO   | Can not get Allure report URL. Please check logs
==================== 1 passed, 4 warnings in 197.46 seconds ====================
INFO:root:Can not get Allure report URL. Please check logs

### CODE REVIEW

**Comment by dks0692** [comment]
@developfast Can you please review

**Comment by rraghav-cisco** [comment]
Hi @dks0692 : where is this function defined: verify_m2o_oversubscribtion_results() ?
When I applied this diff and ran, it gave this error:
`                                                                                                                                     

=========================================================================================================================== ERRORS ===========================================================================================================================
____________________________________________________________________________________ ERROR collecting snappi_tests/multidut/pfc/test_m2o_oversubscribe_lossless_lossy.py _____________________________________________________________________________________
ImportError while importing test module '/data/tests/snappi_tests/multidut/pfc/test_m2o_oversubscribe_lossless_lossy.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
snappi_tests/multidut/pfc/test_m2o_oversubscribe_lossless_lossy.py:11: in <module>
    from tests.snappi_tests.multidut.pfc.files.m2o_oversubscribe_lossless_lossy_helper import (
snappi_tests/multidut/pfc/files/m2o_oversubscribe_lossless_lossy_helper.py:15: in <module>
    from tests.common.snappi_tests.traffic_generation import run_traffic, \
E   ImportError: cannot import name 'verify_m2o_oversubscribtion_results' from 'tests.common.snappi_tests.traffic_generation' (/data/tests/common/snappi_tests/traffic_generation.py)
`

**Comment by rraghav-cisco** [comment]
@selldinesh : thanks for quick reply.

**Comment by rraghav-cisco** [comment]
In our case the test fails since the port statistics includes the other packets from the DUT(PFCXon, BGP, LLDP). The flow statistics shows the correct number of packets, and matches the DUT counters as well:
```
        tx_frames, tx_drop_frames = get_tx_frame_count(duthost, rx_port['peer_port'])
>       pytest_assert(abs(sum_rx - tx_frames)/sum_rx <= rx_frame_count_deviation,
                      "FAIL: DUT counters doesn't match with the total frames received on Rx port, \
                      Deviation of more than {} observed".format(rx_frame_count_deviation))
E       Failed: FAIL: DUT counters doesn't match with the total frames received on Rx port,                   Deviation of more than 0.05 observed

bg_flow_name = 'Background Flow'
criteria   = 'loss'
duthost    = <MultiAsicSonicHost yy39top-lc4>
flag       = {'Background Flow 2 -> 0 Rate:40': 'loss', 'Rate:20': 'no_loss', 'Test Flow': 'no_loss'}
flow_type  = 'Background Flow 2 -> 0 Rate:40'
row        = <snappi.snappi.FlowMetric object at 0x7f57974a2c40>
rows       = <snappi.snappi.FlowMetricIter object at 0x7f579741d300>
rx_frame_count_deviation = 0.05
rx_frames  = 335246290
rx_port    = {'asic_type': 'cisco-8000', 'asic_value': 'asic0', 'card_id': '1', 'ip': '172.27.147.46', ...}
sum_rx     = 1293100696
test_flow_name = 'Test Flow'
tx_drop_frames = 47895472
tx_frames  = 1101529861
```

File: tests/snappi_tests/multidut/pfc/files/m2o_oversubscribe_lossless_lossy_helper.py [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+TEST_FLOW_AGGR_RATE_PERCENT = [40, 20]
+BG_FLOW_NAME = 'Background Flow'
+BG_FLOW_AGGR_RATE_PERCENT = [20, 40]
+DATA_PKT_SIZE = 1024
+DATA_FLOW_DURATION_SEC = 20
+DATA_FLOW_DELAY_SEC = 10
+SNAPPI_POLL_DELAY_SEC = 2
+TOLERANCE_THRESHOLD = 0.05
+
+
```
Comment (vmittal-msft) [inline tests/snappi_tests/multidut/pfc/files/m2o_oversubscribe_lossless_lossy_helper.py:31]: please add comments on how to test is done and expected results.

File: tests/snappi_tests/multidut/pfc/test_m2o_oversubscribe_lossless_lossy.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                                          fanout_graph_facts,           # noqa: F811
+                                          line_card_choice,
+                                          duthosts,
+                                          prio_dscp_map,                # noqa: F811
+                                          lossless_prio_list,           # noqa: F811
+                                          linecard_configuration_set,
+                                          get_multidut_snappi_ports,):  # noqa: F811
+
+    """
+    Run PFC watchdog test under many to one traffic pattern
```
Comment (vmittal-msft) [inline tests/snappi_tests/multidut/pfc/test_m2o_oversubscribe_lossless_lossy.py:32]: please fix comments

**Comment by vmittal-msft** [comment]
please fix test cases

File: tests/snappi_tests/multidut/pfc/files/m2o_oversubscribe_lossless_lossy_helper.py (hunk 1 above)
Comment (dks0692) [reply tests/snappi_tests/multidut/pfc/files/m2o_oversubscribe_lossless_lossy_helper.py:31]: The test data traffic will include two lossless traffic streams, with the SONiC default lossless priorities of 3 and 4, and first stream having a 40% bandwidth from port 1 and second stream having 20% of bandwidth from port 2.
The background traffic will consist of two lossy traffic streams, each with randomly chosen priorities (0..2, 5..7), and one stream to be at 20% from port 1while the other is at 40% bandwidth from port 2. Start the test data traffic from IXIA Tx ports 1; the combined line rate will be 60% of the egress port capacity.
Start the background traffic from IXIA Tx ports 2; the additional increase in bandwidth will be 60%, for a total oversubscription of 120%. After a few seconds, measure the received bandwidth per flow on the IXIA Rx port 3: test data traffic and background traffic with 20% of bandwidth should not lose any packet. Internally the router should be dropping lossy frames from the streams causing the over-subscription (Port 2). There should be PFC pause messages being received at the IXIA transmit port 1 stream 2 causing the over-subscription.

File: tests/snappi_tests/multidut/pfc/test_m2o_oversubscribe_lossless_lossy.py (hunk 2 above)
Comment (dks0692) [reply tests/snappi_tests/multidut/pfc/test_m2o_oversubscribe_lossless_lossy.py]: this has been fixed

File: tests/snappi_tests/multidut/pfc/files/m2o_oversubscribe_lossless_lossy_helper.py (hunk 1 above)
Comment (dks0692) [reply tests/snappi_tests/multidut/pfc/files/m2o_oversubscribe_lossless_lossy_helper.py:31]: The background traffic will include two lossy traffic streams, each with randomly chosen priorities (0..2, 5..7), and each having a 25% bandwidth. The test data traffic will consist of two lossless traffic streams, with the SONiC default lossless priorities of 3 and 4, and each having a 25% bandwidth. Start the background and test traffic from IXIA Tx ports 1 & 2; the combined line rate will be 100% of the egress port capacity (25% line rate bandwidth per stream)After a few seconds, start the PFC pause storm to fully block one or all test priorities at the switch. Ensure that: a. The IXIA Rx port is receiving all lossy streams and any of the lossless streams not marked for PFC pause due to randomization. b. The IXIA Rx port does not receive any of the lossless streams marked to be fully blocked by PFC

File: tests/snappi_tests/multidut/pfc/test_m2o_oversubscribe_lossless_lossy.py (hunk 2 above)
Comment (dks0692) [reply tests/snappi_tests/multidut/pfc/test_m2o_oversubscribe_lossless_lossy.py]: fixed

**Comment by selldinesh** [comment]
@yxieca @wangxin please help merge.

**Review by vmittal-msft — APPROVED** [review]

**Comment by vmittal-msft** [comment]
@developfast does this look good from your side ?

**Comment by vmittal-msft** [comment]
@selldinesh please help fix test failures.

**Comment by vmittal-msft** [comment]
@yxieca @wangxin all tests passing now. please help merge.

**Pull request merged**